presto-main/src/main/java/com/facebook/presto/operator/SortedPositionLinks.java

```diff
@@ -192,53 +193,54 @@ public int next(int position, int probePosition, Page allProbeChannelsPage)
     @Override
     public int start(int startingPosition, int probePosition, Page allProbeChannelsPage)
     {
+        int[] links = sortedPositionLinks[startingPosition];
+        if (links == null) {
+            return -1;
+        }
+        int currentStartOffset = 0;
         for (JoinFilterFunction searchFunction : searchFunctions) {
-            startingPosition = findStartPositionForFunction(searchFunction, startingPosition, probePosition, allProbeChannelsPage);
+            currentStartOffset = findStartPositionForFunction(searchFunction, links, currentStartOffset, probePosition, allProbeChannelsPage);
             // return as soon as a mismatch is found, since we are handling only AND predicates (conjuncts)
-            if (startingPosition == -1) {
+            if (currentStartOffset == -1) {
                 return -1;
             }
         }
-        return startingPosition;
+        return links[currentStartOffset];
     }
 
-    private int findStartPositionForFunction(JoinFilterFunction searchFunction, int startingPosition, int probePosition, Page allProbeChannelsPage)
+    private int findStartPositionForFunction(JoinFilterFunction searchFunction, int[] links, int startOffset, int probePosition, Page allProbeChannelsPage)
     {
         // check if filtering function to startingPosition
-        if (applySearchFunction(searchFunction, startingPosition, probePosition, allProbeChannelsPage)) {
-            return startingPosition;
-        }
-
-        if (sortedPositionLinks[startingPosition] == null) {
-            return -1;
+        if (applySearchFunction(searchFunction, links, startOffset, probePosition, allProbeChannelsPage)) {
+            return startOffset;
         }
 
-        int left = 0;
-        int right = sortedPositionLinks[startingPosition].length - 1;
+        int left = startOffset;
+        int right = links.length - 1;
 
         // do a binary search for the first position for which filter function applies
-        int offset = lowerBound(searchFunction, startingPosition, left, right, probePosition, allProbeChannelsPage);
+        int offset = lowerBound(searchFunction, links, left, right, probePosition, allProbeChannelsPage);
         if (offset < 0) {
             return -1;
         }
-        if (!applySearchFunction(searchFunction, startingPosition, offset, probePosition, allProbeChannelsPage)) {
+        if (!applySearchFunction(searchFunction, links, offset, probePosition, allProbeChannelsPage)) {
```

Comment:
Now i dont understand what this `if` does here. Looks like it's redundant and `lowerBound` could cover it. (no change required, but you can think about it..)